### PR TITLE
Add an Orthogonal Projection option to the Camera Gem

### PR DESCRIPTION
**Pull request by NicholasVanSickle**
This adds a check-box to opt into an ortho projection along with a half-width parameter to adjust the size of the visible area. Includes some light tweaks to ensure debug rendering looks OK and that we generate a correct camera state for these non-perspective views.

Known issue: while in "Be this camera" mode in the Editor using an ortho projection manipulators aren't working correctly. This appears to be a downstream issue with CameraState consumers not actually checking the ortho flag.

orthocamera

Signed-off-by: nvsickle <nvsickle@amazon.com>

### CODE REVIEW

File: Code/Framework/AzFramework/AzFramework/Components/CameraBus.h [hunk 1]
```diff
@@ -63,6 +63,13 @@ namespace Camera
         //! @return The camera frustum's height
         virtual float GetFrustumHeight() = 0;
 
+        //! Gets whether or not the camera is using an orthographic projection.
+        //! @return True if the camera is using an orthographic projecetion, or false if the camera is a using a perspective projection.
```
Comment (AMZN-daimini) [inline Code/Framework/AzFramework/AzFramework/Components/CameraBus.h:67]: nit: projec _e_ tion
nit: if the camera is _a_ using a perspective projection.

File: Gems/Camera/Code/Source/CameraComponentController.cpp [hunk 2]
```diff
@@ -42,25 +44,33 @@ namespace Camera
...
                             "Make active camera on activation?", "If true, this camera will become the active render camera when it activates")
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &CameraComponentConfig::m_orthographic, "Orthographic",
+                        "If set, this camera will use an orthographic projection instead of a perspective one. Objects will appear as the same size, regardless of distance from the camera.")
+                        ->Attribute(AZ::Edit::Attributes::ChangeNotify, AZ::Edit::PropertyRefreshLevels::EntireTree)
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &CameraComponentConfig::m_orthographicHalfWidth, "Orthographic Half-width", "The half-width used to calculate the orthographic projection. The height will be determined by the aspect ratio.")
+                        ->Attribute(AZ::Edit::Attributes::Visibility, &CameraComponentConfig::GetOrthographicParameterVisibility)
+                        ->Attribute(AZ::Edit::Attributes::Min, 0.001f)
+                        ->Attribute(AZ::Edit::Attributes::ChangeNotify, AZ::Edit::PropertyRefreshLevels::ValuesOnly)
```
Comment (AMZN-daimini) [inline Gems/Camera/Code/Source/CameraComponentController.cpp:53]: nit: Could this use a custom step?

**Review by AMZN-daimini — APPROVED** [review]

File: Gems/Camera/Code/Source/CameraComponentController.cpp (hunk 2 above)
Comment (NicholasVanSickle) [reply Gems/Camera/Code/Source/CameraComponentController.cpp:53]: The default step feels appropriate from my initial playing with it and I'm not sure exactly what the custom step would ideally be (I suspect more on the fine tuning side than the broad strokes side) so I'm inclined to leave it be for now

File: Gems/Camera/Code/Source/EditorCameraComponent.cpp [hunk 3]
```diff
@@ -155,6 +158,42 @@ namespace Camera
...
+        const CameraComponentConfig& config = m_controller.GetConfiguration();
+        AZ::RPI::ViewportContextPtr viewportContext = m_controller.GetViewportContext();
+        AZ::RPI::ViewPtr view = m_controller.GetView();
+
+        if (viewportContext == nullptr || view == nullptr)
+        {
+            return false;
+        }
+
+        AzFramework::SetCameraTransform(cameraState, view->GetCameraTransform());
```
Comment (moraaar) [inline Gems/Camera/Code/Source/EditorCameraComponent.cpp:172]: It doesn't look right that a getter function (`GetCameraState`) calls a a setter function to set the camera transform. Why is this needed?

File: Gems/Camera/Code/Source/EditorCameraComponent.cpp [hunk 4]
```diff
@@ -201,22 +240,42 @@ namespace Camera
...
         farPoints[2] = AZ::Vector3(-width, distance, -height);
         farPoints[3] = AZ::Vector3( width, distance, -height);
 
-        AZ::Vector3 start(0, 0, 0);
         AZ::Vector3 nearPoints[4];
-        nearPoints[0] = farPoints[0].GetNormalizedSafe() * config.m_nearClipDistance;
-        nearPoints[1] = farPoints[1].GetNormalizedSafe() * config.m_nearClipDistance;
-        nearPoints[2] = farPoints[2].GetNormalizedSafe() * config.m_nearClipDistance;
-        nearPoints[3] = farPoints[3].GetNormalizedSafe() * config.m_nearClipDistance;
+        if (config.m_orthographic)
```
Comment (moraaar) [inline Gems/Camera/Code/Source/EditorCameraComponent.cpp:265]: Adding a option to the camera component to choose orthographic camera vs perspective camera is a good approach, and for now it's great. For long term this approach should be run through UX team to gather feedback. Maybe they'd have arguments for a different approach, like for example providing two different camera components.

File: Gems/Camera/Code/Source/CameraComponentController.cpp [hunk 5]
```diff
@@ -423,30 +475,37 @@ namespace Camera
...
+            }
+            else
+            {
+                AZ::MakePerspectiveFovMatrixRH(viewToClipMatrix,
+                    AZ::DegToRad(m_config.m_fov),
+                    aspectRatio,
+                    m_config.m_nearClipDistance,
+                    m_config.m_farClipDistance,
+                    true);
+            }
```
Comment (moraaar) [inline Gems/Camera/Code/Source/CameraComponentController.cpp:508]: question: just double checking the near/far arguments passed are correct. They have different order, in `MakeOrthographicMatrixRH` far is passed first, then near, and in `MakePerspectiveFovMatrixRH` near is passed first, then far.
Is that correct? If it is then we should change the functions to be consistent and avoid potential errors in the future.

File: Gems/Camera/Code/Source/EditorCameraComponent.cpp (hunk 3 above)
Comment (NicholasVanSickle) [reply Gems/Camera/Code/Source/EditorCameraComponent.cpp:172]: The method name is a bit misleading in this context; this is just applying the camera transform to `cameraState` (internally it just decomposes the transform into forward/up/left normals plus a translation)

File: Gems/Camera/Code/Source/EditorCameraComponent.cpp (hunk 4 above)
Comment (NicholasVanSickle) [reply Gems/Camera/Code/Source/EditorCameraComponent.cpp:265]: Agreed! This is just a quick fix for now

File: Gems/Camera/Code/Source/CameraComponentController.cpp (hunk 5 above)
Comment (NicholasVanSickle) [reply Gems/Camera/Code/Source/CameraComponentController.cpp]: Ah yeah `MakeOrthographicMatrixRH` didn't have a `reverseDepth` parameter like the other helpers - I went ahead and added one to make this less confusing, as it's very easy to get wrong (including in the gif recorded above!)

Comment (moraaar) [reply Gems/Camera/Code/Source/CameraComponentController.cpp]: That's great, thanks!

**Review by alexmontAmazon — APPROVED** [review]
